### ISSUE TITLE
Endless Sign In Loop

### ISSUE THREAD

**Issue by STLCardsFan83** [miscord/miscord#434]
Looking for a way to have the bot retry Facebook sign in after Facebook requests sign in verification. 

Current state (in layman’s terms) when Facebook requires verification of login attempt:

Bot logs in, Facebook says hey please verify that you logged in, I say yes that was me, it says cool.  The bot then says but Facebook told me no so you have to restart me.  I restart the bot, Facebook says hey please verify that you logged in..... and so the loop goes.

**Comment by glensc** [comment]
So, happened me too. first error is different:

```
 SUCCESS  22:17:37 messenger › login ›  Logged in to Facebook
 SUCCESS  22:17:37 login ›  Logged in
This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason:
AssertionError [ERR_ASSERTION]: Method 'GET' already declared for route '/connections/'
    at Router._insert (/home/users/glen/scm/miscord/node_modules/find-my-way/index.js:286:9)
    at Router._on (/home/users/glen/scm/miscord/node_modules/find-my-way/index.js:188:8)
    at Router.on (/home/users/glen/scm/miscord/node_modules/find-my-way/index.js:73:8)
    at Object.afterRouteAdded (/home/users/glen/scm/miscord/node_modules/fastify/lib/route.js:232:16)
    at /home/users/glen/scm/miscord/node_modules/fastify/lib/route.js:163:31
    at Object._encapsulateThreeParam (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:468:7)
    at Boot.timeoutCall (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:379:5)
    at Boot.callWithCbOrNextTick (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:360:19)
    at Boot._after (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:235:26)
    at Plugin.exec (/home/users/glen/scm/miscord/node_modules/avvio/plugin.js:89:17)
  ERROR   22:17:38 errorHandler ›  AssertionError [ERR_ASSERTION]: Method 'GET' already declared for route '/connections/'
  ERROR   22:17:38 errorHandler ›      at Router._insert (/home/users/glen/scm/miscord/node_modules/find-my-way/index.js:286:9)
  ERROR   22:17:38 errorHandler ›      at Router._on (/home/users/glen/scm/miscord/node_modules/find-my-way/index.js:188:8)
  ERROR   22:17:38 errorHandler ›      at Router.on (/home/users/glen/scm/miscord/node_modules/find-my-way/index.js:73:8)
  ERROR   22:17:38 errorHandler ›      at Object.afterRouteAdded (/home/users/glen/scm/miscord/node_modules/fastify/lib/route.js:232:16)
  ERROR   22:17:38 errorHandler ›      at /home/users/glen/scm/miscord/node_modules/fastify/lib/route.js:163:31
  ERROR   22:17:38 errorHandler ›      at Object._encapsulateThreeParam (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:468:7)
  ERROR   22:17:38 errorHandler ›      at Boot.timeoutCall (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:379:5)
  ERROR   22:17:38 errorHandler ›      at Boot.callWithCbOrNextTick (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:360:19)
  ERROR   22:17:38 errorHandler ›      at Boot._after (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:235:26)
  ERROR   22:17:38 errorHandler ›      at Plugin.exec (/home/users/glen/scm/miscord/node_modules/avvio/plugin.js:89:17) {
  ERROR   22:17:38 errorHandler ›    generatedMessage: false,
  ERROR   22:17:38 errorHandler ›    code: 'ERR_ASSERTION',
  ERROR   22:17:38 errorHandler ›    actual: false,
  ERROR   22:17:38 errorHandler ›    expected: true,
  ERROR   22:17:38 errorHandler ›    operator: '=='
  ERROR   22:17:38 errorHandler ›  }
  INFO    22:17:38  ›  Worker process 31485 died (2, null).
```

after auto-restart the error keeps repeating:

```
  START   22:18:26 messenger › login ›  Logging in to Facebook...
  WARN    22:18:26 messenger › login ›  Could not log into Facebook
  ERROR   22:18:26 errorHandler ›  AuthAPIError: User must verify their account on www.facebook.com (405)
  ERROR   22:18:26 errorHandler ›      at /home/users/glen/scm/miscord/node_modules/libfb/dist/http/HttpApi.js:26:35
  ERROR   22:18:26 errorHandler ›      at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  ERROR   22:18:26 errorHandler ›    code: 405,
  ERROR   22:18:26 errorHandler ›    errorData: {
  ERROR   22:18:26 errorHandler ›      url: 'https://www.facebook.com/checkpoint/start/?ip=*omittted*&cookie=*omittted*,
  ERROR   22:18:26 errorHandler ›      flow_id: *omittted*,
  ERROR   22:18:26 errorHandler ›      uid: *omittted*,
  ERROR   22:18:26 errorHandler ›      show_native_checkpoints: false,
  ERROR   22:18:26 errorHandler ›      start_internal_webview_from_url: true,
  ERROR   22:18:26 errorHandler ›      error_title: 'Account Temporarily Unavailable',
  ERROR   22:18:26 errorHandler ›      error_message: 'Your account is temporarily unavailable. Regain access by updating the app or logging in from a mobile or web browser.'
  ERROR   22:18:26 errorHandler ›    },
  ERROR   22:18:26 errorHandler ›    requestArgs: {
  ERROR   22:18:26 errorHandler ›      method: 'auth.login',
  ERROR   22:18:26 errorHandler ›      api_key: '*omittted*',
  ERROR   22:18:26 errorHandler ›      device_id: 'undefined',
  ERROR   22:18:26 errorHandler ›      email: '*omittted*',
  ERROR   22:18:26 errorHandler ›      fb_api_req_friendly_name: 'authenticate',
  ERROR   22:18:26 errorHandler ›      format: 'json',
  ERROR   22:18:26 errorHandler ›      locale: 'en_EN',
  ERROR   22:18:26 errorHandler ›      password: '--sanitized--',
  ERROR   22:18:26 errorHandler ›      sig: '*omittted*'
  ERROR   22:18:26 errorHandler ›    }
  ERROR   22:18:26 errorHandler ›  }
  ERROR   22:18:26 errorHandler ›
  ERROR   22:18:26 errorHandler ›  Verify your account here: https://www.facebook.com/checkpoint/start/?ip=*omittted*&cookie=*omittted*
  ERROR   22:18:26 errorHandler ›  Account: *omittted*
  ERROR   22:18:26 errorHandler ›
```

and unfortunately, app restarted itself, repeating those errors, making the FB account locked out, so visiting the verify link has no help:

> Your account has been temporarily locked

**Comment by glensc** [comment]
here's log after recovering fb account. but seems it deleted all app passwords, i was using app password for this app.

anyway, i think the app must not restart and retry, as such behaviour will result lock out of account.

```
➔ ./miscord -d .
 SUCCESS  22:37:06  ›  Worker process 2654 started.
  START   22:37:06  ›  Compressing old logs...
 SUCCESS  22:37:06  ›  All old logs compressed, starting Miscord
  INFO    22:37:06 FileConfig ›  Loading config from /home/users/glen/scm/miscord/config.json
  START   22:37:06  ›  Launching Miscord v5.0.4
  START   22:37:06 login ›  Logging in...
  INFO    22:37:06 login › logLevel 'info'
  START   22:37:06 discord › login ›  Logging in to Discord...
 SUCCESS  22:37:06 discord › login ›  Logged in to Discord
  START   22:37:06 messenger › login ›  Logging in to Facebook...
  WARN    22:37:12 messenger › login ›  Could not log into Facebook
  ERROR   22:37:12 errorHandler ›  AuthAPIError: Invalid username or password (401)
  ERROR   22:37:12 errorHandler ›      at /home/users/glen/scm/miscord/node_modules/libfb/dist/http/HttpApi.js:26:35
  ERROR   22:37:12 errorHandler ›      at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  ERROR   22:37:12 errorHandler ›    code: 401,
  ERROR   22:37:12 errorHandler ›    errorData: { error_title: 'Incorrect Password', error_message: 'The password you entered is incorrect. Please try again.' },
  ERROR   22:37:12 errorHandler ›    requestArgs: {
  ERROR   22:37:12 errorHandler ›      method: 'auth.login',
  ERROR   22:37:12 errorHandler ›      api_key: '*omitted*',
  ERROR   22:37:12 errorHandler ›      device_id: 'undefined',
  ERROR   22:37:12 errorHandler ›      email: '*omitted*',
  ERROR   22:37:12 errorHandler ›      fb_api_req_friendly_name: 'authenticate',
  ERROR   22:37:12 errorHandler ›      format: 'json',
  ERROR   22:37:12 errorHandler ›      locale: 'en_EN',
  ERROR   22:37:12 errorHandler ›      password: '--sanitized--',
  ERROR   22:37:12 errorHandler ›      sig: '*omitted*'
  ERROR   22:37:12 errorHandler ›    }
  ERROR   22:37:12 errorHandler ›  }
  ERROR   22:37:12 errorHandler ›
  ERROR   22:37:12 errorHandler ›  Couldn't log into Facebook.
  ERROR   22:37:12 errorHandler ›  Check your password or preferrably, use an app password:
  ERROR   22:37:12 errorHandler ›  http://facebook.com/settings?tab=security&section=per_app_passwords&view
  ERROR   22:37:12 errorHandler ›
 SUCCESS  22:37:12 login ›  Logged in
This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason:
AssertionError [ERR_ASSERTION]: Method 'GET' already declared for route '/connections/'
    at Router._insert (/home/users/glen/scm/miscord/node_modules/find-my-way/index.js:286:9)
    at Router._on (/home/users/glen/scm/miscord/node_modules/find-my-way/index.js:188:8)
    at Router.on (/home/users/glen/scm/miscord/node_modules/find-my-way/index.js:73:8)
    at Object.afterRouteAdded (/home/users/glen/scm/miscord/node_modules/fastify/lib/route.js:232:16)
    at /home/users/glen/scm/miscord/node_modules/fastify/lib/route.js:163:31
    at Object._encapsulateThreeParam (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:468:7)
    at Boot.timeoutCall (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:379:5)
    at Boot.callWithCbOrNextTick (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:360:19)
    at Boot._after (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:235:26)
    at Plugin.exec (/home/users/glen/scm/miscord/node_modules/avvio/plugin.js:89:17)
  ERROR   22:37:13 errorHandler ›  AssertionError [ERR_ASSERTION]: Method 'GET' already declared for route '/connections/'
  ERROR   22:37:13 errorHandler ›      at Router._insert (/home/users/glen/scm/miscord/node_modules/find-my-way/index.js:286:9)
  ERROR   22:37:13 errorHandler ›      at Router._on (/home/users/glen/scm/miscord/node_modules/find-my-way/index.js:188:8)
  ERROR   22:37:13 errorHandler ›      at Router.on (/home/users/glen/scm/miscord/node_modules/find-my-way/index.js:73:8)
  ERROR   22:37:13 errorHandler ›      at Object.afterRouteAdded (/home/users/glen/scm/miscord/node_modules/fastify/lib/route.js:232:16)
  ERROR   22:37:13 errorHandler ›      at /home/users/glen/scm/miscord/node_modules/fastify/lib/route.js:163:31
  ERROR   22:37:13 errorHandler ›      at Object._encapsulateThreeParam (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:468:7)
  ERROR   22:37:13 errorHandler ›      at Boot.timeoutCall (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:379:5)
  ERROR   22:37:13 errorHandler ›      at Boot.callWithCbOrNextTick (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:360:19)
  ERROR   22:37:13 errorHandler ›      at Boot._after (/home/users/glen/scm/miscord/node_modules/avvio/boot.js:235:26)
  ERROR   22:37:13 errorHandler ›      at Plugin.exec (/home/users/glen/scm/miscord/node_modules/avvio/plugin.js:89:17) {
  ERROR   22:37:13 errorHandler ›    generatedMessage: false,
  ERROR   22:37:13 errorHandler ›    code: 'ERR_ASSERTION',
  ERROR   22:37:13 errorHandler ›    actual: false,
  ERROR   22:37:13 errorHandler ›    expected: true,
  ERROR   22:37:13 errorHandler ›    operator: '=='
  ERROR   22:37:13 errorHandler ›  }
  INFO    22:37:13  ›  Worker process 2654 died (2, null).
^C
```

**Comment by glensc** [comment]
Oddly, the error above is not related to FB login at all. but setting up routes, those errors go away if I remove "api" block from config.json and app logs in successfully. some initialization block invoked more than once?

**Comment by glensc** [comment]
I've filed a feature request to at least minimize impact when the app goes into such a loop, so don't have to go through the process of recovering the FB account:
- https://github.com/miscord/miscord/issues/472

**Comment by jarp0l** [comment]
```
...
AssertionError [ERR_ASSERTION]: Method 'GET' already declared for route '/connections/'
...
...
```
*^from glensc's first comment above*

I don't know much of this, but just wanted to report that the `AssertionError` did not arise when I ran miscord in [repl.it](https://repl.it), even with the "api" block present in config.json. And I was able to view the dashboard without any error.

Before running miscord in repl.it I cloned the dev branch there, and compiled and ran it. As far as I remember the only change I did was in [line#89](https://github.com/miscord/miscord/blob/dev/src/config/FileConfig.ts#L89) of FileConfig.ts (FileConfig.js), where I returned the same thing as in case of linux because it was throwing a permission related error.

**Comment by r1bnc** [comment]
Same problem even with app password
